src/Databases/DatabaseReplicated.cpp

```diff
@@ -0,0 +1,288 @@
+    // that no longer will be in use by current replicas or
+    // new coming ones.
+    // Each registered replica has its state in ZooKeeper.
+    // Therefore, snapshots and logged queries that are less
+    // than a least advanced replica are removed.
+    // It does not interfere with a new coming replica
+    // metadata loading from snapshot
+    // because the replica will use the latest snapshot available
+    // and this snapshot will set the last executed log query
+    // to a greater one than the least advanced current replica.
+    current_zookeeper = getZooKeeper();
+    Strings replica_states = current_zookeeper->getChildren(zookeeper_path + "/replicas");
+    auto least_advanced = std::min_element(replica_states.begin(), replica_states.end());
+    Strings snapshots = current_zookeeper->getChildren(zookeeper_path + "/snapshots");
+    
+    if (snapshots.size() < 2) {
+        return;
+    }
+
+    std::sort(snapshots.begin(), snapshots.end());
+    auto still_useful = std::lower_bound(snapshots.begin(), snapshots.end(), *least_advanced);
+    snapshots.erase(still_useful, snapshots.end());
+    for (const String & snapshot : snapshots) {
+        current_zookeeper->tryRemoveRecursive(zookeeper_path + "/snapshots/" + snapshot);
+    }
+
+    Strings log_entry_names = current_zookeeper->getChildren(zookeeper_path + "/log");
+    std::sort(log_entry_names.begin(), log_entry_names.end());
+    auto still_useful_log = std::upper_bound(log_entry_names.begin(), log_entry_names.end(), *still_useful);
+    log_entry_names.erase(still_useful_log, log_entry_names.end());
+    for (const String & log_entry_name : log_entry_names) {
+        String log_entry_path = zookeeper_path + "/log/" + log_entry_name;
+        current_zookeeper->tryRemove(log_entry_path);
+    }
+}
+
+void DatabaseReplicated::runBackgroundLogExecutor() {
+    if (last_executed_log_entry == "") {
+        loadMetadataFromSnapshot();
+    }
+
+    current_zookeeper = getZooKeeper();
+    Strings log_entry_names = current_zookeeper->getChildren(zookeeper_path + "/log");
+
+    std::sort(log_entry_names.begin(), log_entry_names.end());
+    auto newest_entry_it = std::upper_bound(log_entry_names.begin(), log_entry_names.end(), last_executed_log_entry);
+
+    log_entry_names.erase(log_entry_names.begin(), newest_entry_it);
+
+    for (const String & log_entry_name : log_entry_names) {
+        String log_entry_path = zookeeper_path + "/log/" + log_entry_name;
+        executeFromZK(log_entry_path);
+        last_executed_log_entry = log_entry_name;
+        writeLastExecutedToDiskAndZK();
+
+        int log_n = parse<int>(log_entry_name.substr(4));
+        int last_log_n = parse<int>(log_entry_names.back().substr(4));
+
+        // The third condition gurantees at most one snapshot creation per batch
+        if (log_n > 0 && snapshot_period > 0 && (last_log_n - log_n) / snapshot_period == 0 && log_n % snapshot_period == 0) {
```

Comment:
Could you please explain what does the third condition mean and why the fourth it not enough?

Reply:
Looks like the second and the third conditions can be simplified to `last_log_n - log_n < snapshot_period`.

Reply:
`snapshot_period > 0` means that snapshotting is enabled at all.
`log_n % snapshot_period == 0` means that one snapshot is created per `snapshot_period` logged queries.
`(last_log_n - log_n) / snapshot_period == 0` is a bit confusing. Let me give an example.
`snapshot_period` is set to 2 (one snapshot per two queries), we got a range of logs from 1 to 19, and therefore `last_log_n` is 19. The idea is that for the given replica even though snapshot period is 2, there's no point in snapshots
2, 4, 6, and so on. Replica is not much advanced since it got so many new logs and no one will get anything from snapshots of old states. The condition `(last_log_n - log_n) / snapshot_period == 0` ensures that this replica will only create a snapshot for `log_18` and `log_19`. However, in this case snapshot period is 2, and only `log_18` will trigger `createSnapshot`. The third condition is only useful when there's a log of new log queries compared to `snapshot_period`.
At least in my testing environment `createSnapshot` takes some time and this hack makes the database engine more rapid for this specific case.